File: library/src/main/java/me/proxer/library/api/ConferenceAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.Json;
import lombok.AccessLevel;
import lombok.Getter;
import me.proxer.library.entitiy.messenger.Conference;
import org.jetbrains.annotations.NotNull;

import java.util.Date;

/**
 * @author Ruben Gees
 */
class ConferenceAdapter {

    @FromJson
    Conference fromJson(IntermediateConference json) {
        return new Conference(json.id, json.topic, json.customTopic, json.participantAmount, json.getImage(),
                json.getImageType(), json.group, json.read, json.date, json.unreadMessageAmount,
                json.lastReadMessageId);
    }

    static class IntermediateConference {

        private static final String IMAGE_DELIMITER = ":";
        private static final String EMPTY_RESULT = "";

        @Json(name = "id")
        String id;

        @Json(name = "topic")
        String topic;

        @Json(name = "topic_custom")
        String customTopic;

        @Json(name = "count")
        int participantAmount;

        @Json(name = "group")
        boolean group;

        @Json(name = "read")
        boolean read;

        @Json(name = "timestamp_end")
        Date date;

        @Json(name = "read_count")
        int unreadMessageAmount;

        @Json(name = "read_mid")
        String lastReadMessageId;

        @Getter(AccessLevel.NONE)
        @Json(name = "image")
        String image;

        @NotNull
        String getImage() {
            final int delimiterIndex = image.indexOf(IMAGE_DELIMITER);

            if (delimiterIndex < 0) {
                return EMPTY_RESULT;
            } else {
                return image.substring(0, delimiterIndex);
            }
        }

        @NotNull
        String getImageType() {
            final int delimiterIndex = image.indexOf(IMAGE_DELIMITER);

            if (delimiterIndex < 0) {
                return EMPTY_RESULT;
            } else {
                return image.substring(delimiterIndex + 1);
            }
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/api/ConferenceAdapter.kt
package me.proxer.library.api

import com.squareup.moshi.FromJson
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.proxer.library.entity.messenger.Conference
import java.util.Date

/**
 * @author Ruben Gees
 */
internal class ConferenceAdapter {

    private companion object {
        private const val IMAGE_DELIMITER = ":"
    }

    @FromJson
    fun fromJson(json: IntermediateConference): Conference {
        return Conference(
            json.id, json.topic, json.customTopic, json.participantAmount, json.image, json.imageType,
            json.isGroup, json.isRead, json.date, json.unreadMessageAmount, json.lastReadMessageId
        )
    }

    @JsonClass(generateAdapter = true)
    internal class IntermediateConference(
        @Json(name = "id") val id: String,
        @Json(name = "topic") val topic: String,
        @Json(name = "topic_custom") val customTopic: String,
        @Json(name = "count") val participantAmount: Int,
        @Json(name = "group") val isGroup: Boolean,
        @Json(name = "read") val isRead: Boolean,
        @Json(name = "timestamp_end") val date: Date,
        @Json(name = "read_count") val unreadMessageAmount: Int,
        @Json(name = "read_mid") val lastReadMessageId: String,
        @Json(name = "image") val rawImage: String
    ) {

        val imageType = rawImage.substringBefore(IMAGE_DELIMITER, "")
        val image = rawImage.substringAfter(IMAGE_DELIMITER, "")
    }
}
